### PR TITLE
List: Parameterize generic type for item with any as default

### PR DESCRIPTION
**Pull request by KevinTCoughlin**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [x] Include a change request file using `$ npm run change`

#### Description of changes

I've found working with `List` and components that wrap it difficult as an end-user. This is partially because the backing `items` Array is typed as `any[]` and therefore type-safety is lost for TypeScript developers.

Last week on Shield, I watched a developer run into problems during a debug session as a result of this lack of type safety and thought it was worth fixing **IFF** backwards-compatibility could be maintained.

To reduce this barrier, I'm proposing that we parameterize the generic type for `List`'s `items` as `T` which defaults to `any`.

This _should_ maintain backwards compatibility as `item` and `items` will continue to be typed as `any` if a parameterized type is omitted. If the type is provided, then the intellisense and other goodness should then be provided.

**Note:** As part of our next major release we should look to remove the need for `any` as the default type.

#### IDE Output

Here is a brief comparison of Code's intellisense _before_ and _after_ this change for the code snippet below:

**Usage**:

```tsx
export interface IListBasicExampleProps {
  items: IExampleItem[];
}

<List items={items} onRenderCell={this._onRenderCell} />
```

**Before**

`onRenderCell` prop:

```ts
(JSX attribute) onRenderCell?: ((item?: any, index?: number | undefined, isScrolling?: boolean | undefined) => React.ReactNode) | undefined
Method to call when trying to render an item.
```

`items` prop:

```ts
(JSX attribute) items?: any[] | undefined
Items to render.
```

`List` class definition

```ts
(alias) class List
import List
```

**After**

`onRenderCell` prop:
```ts
(JSX attribute) onRenderCell?: ((item?: IExampleItem | undefined, index?: number | undefined, isScrolling?: boolean | undefined) => React.ReactNode) | undefined
Method to call when trying to render an item.
```

`items` prop:

```ts
(JSX attribute) items?: IExampleItem[] | undefined
Items to render.
```

`List` class definition

```ts
(alias) class List<T = any>
import List
```

#### Focus areas to test

- CI should pass
- API file modifications should maintain backwards compatibility

Related to: #5557 

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/8465)

### CODE REVIEW

**Comment by KevinTCoughlin** [comment]
My **primary** concern here is maintaining backwards-compatibility. I believe this change achieves that. However, if it does not then I am fine with targeting `7.0` branch.

File: apps/todo-app/src/components/TodoTabs.tsx [hunk 1]
```diff
@@ -61,7 +61,10 @@ export default class TodoTabs extends React.Component<ITodoTabsProps, {}> {
     return ev.which === KeyCodes.right;
   };
 
-  private _onRenderTodoItem(item: ITodoItem): React.ReactElement<ITodoItemProps> {
```
Comment (KevinTCoughlin) [inline apps/todo-app/src/components/TodoTabs.tsx]: `item` is not guaranteed to be defined _before_ or _after_ this change.

**Comment by ecraig12345** [comment]
Agreed that with T = any, you'll probably be safe for compatibility. Someone else correct me if I'm wrong.

File: packages/office-ui-fabric-react/src/components/List/List.tsx [hunk 2]
```diff
@@ -25,21 +25,21 @@ const DEFAULT_RENDERED_WINDOWS_AHEAD = 2;
 const PAGE_KEY_PREFIX = 'page-';
 const SPACER_KEY_PREFIX = 'spacer-';
 
-export interface IListState {
-  pages?: IPage[];
+export interface IListState<T = any> {
```
Comment (cliffkoh) [inline packages/office-ui-fabric-react/src/components/List/List.tsx]: ><T = any> [](start = 27, length = 9)

:+1:

**Comment by KevinTCoughlin** [comment]
For anyone that may land here due to compiler errors, the solution is likely as follows:

If you have previously defined your List's `onRenderCell` method as:

```tsx
onRenderCell?: (item: Foo, index?: number, isScrolling?: boolean) => React.ReactNode;
```

Where item of type `Foo` is not marked as an _optional_ parameter then you will need to mark it as such as it [has been defined in IListProps](https://github.com/OfficeDev/office-ui-fabric-react/blob/3029f67059a367f6077731a648e19b13c38de492/packages/office-ui-fabric-react/src/components/List/List.types.ts#L71). The modified signature would look like:

```tsx
onRenderCell?: (item?: Foo, index?: number, isScrolling?: boolean) => React.ReactNode;
```

Also, in the body of the function you may want to guard against the case where item is `undefined`.

This bug was previously masked by the usage of `any`.

**Pull request merged**